cosim/emulator/src/spike_event.cc

```diff
@@ -0,0 +1,248 @@
+  }
+  // since log_mem_read doesn't record mem data, we need to load manually
+  for (auto mem_read: state->log_mem_read) {
+    uint64_t address = std::get<0>(mem_read);
+    if (address != target_mem) {
+      LOG(FATAL_S) << fmt::format("Error! spike detect mem_read at= {:08X}; target mem = {:08X}", address, target_mem);
+    }
+    // Byte size_bytes
+    uint8_t size_by_byte = std::get<2>(mem_read);
+    uint64_t value = 0;
+    // record mem target
+    for (int i = 0; i < size_by_byte; ++i) {
+      value += (uint64_t) impl->load(address + i) << (i * 8);
+    }
+    LOG(INFO)
+        << fmt::format("spike detect mem read {:08X} on mem:{:08X} with size={}byte", value, address, size_by_byte);
+    mem_access_record.all_reads[address] = {.size_by_byte = size_by_byte, .val = value};
+  }
+
+  // record root page table
+  if (satp_mode == 0x8 && block.addr == -1) {
+    uint64_t root_addr = satp_ppn << 12;
+    for (int i = 0; i < 8; i++) {
+      uint64_t data = 0;
+      //scan 8 bytes to data
+      for (int j = 0; j < 8; ++j) {
+        data += (uint64_t) impl->load(root_addr + j + i * 8) << (j * 8);
+      }
+      block.blocks[i] = data;
+      block.addr = root_addr;
+      block.remaining = true;
+    }
+  }
+
+  state->log_reg_write.clear();
+  state->log_mem_read.clear();
+  state->log_mem_write.clear();
+}
+
+SpikeEvent::SpikeEvent(processor_t &proc, insn_fetch_t &fetch, VBridgeImpl *impl) : proc(proc), impl(impl) {
+  auto &xr = proc.get_state()->XPR;
+  pc = proc.get_state()->pc;
+  inst_bits = fetch.insn.bits();
+  target_mem = -1;
+  is_committed = false;// j insn should be committed immediately cause it doesn't have wb stage.
+
+  // extension depending parameter
+  is_compress = false;
+  if (fetch.insn.length() == 2) {
+    is_compress = true;
+  }
+  if (!is_compress) {
+    rs1_bits = xr[fetch.insn.rs1()];
+    rs2_bits = xr[fetch.insn.rs2()];
+    rd_idx = fetch.insn.rd();
+    opcode = clip(inst_bits, 0, 6);
+    uint8_t func7 = (inst_bits & 0xFE000000) >> 25;
+    // for j insn for x0;
+    if (opcode == 0b1101111 && rd_idx == 0) {
+      is_committed = true;
+    }
+    // for integer/double load/store
+    is_load = (opcode == 0b11) || (opcode == 0b0000111);
+    is_store = opcode == 0b100011 || (opcode == 0b0100111);
+    is_amo = opcode == 0b0101111;
+
+    is_mutiCycle = (opcode == 0b0110011)&&(func7 == 0b1);
+
+    if (is_load) {
+      target_mem = rs1_bits + fetch.insn.i_imm();
+    }
+    if (is_store) target_mem = rs1_bits + fetch.insn.s_imm();
+    if (is_amo) target_mem = rs1_bits;
+  } else {
+    uint8_t op = inst_bits & 0b11;
+    uint8_t func3 = (inst_bits & 0xE000) >> 13;
+    uint64_t rs1s_bits = xr[fetch.insn.rvc_rs1s()];
+    uint64_t sp_bits = xr[2];
+    switch (op) {
+      case 0:
+        rd_idx = 8 + ((inst_bits & 0b11100) >> 2);// todo: rd_idx for store insn?
+        if (func3 >= 5) {
+          is_store = true;
+          switch (func3) {
```

Comment:
not sure if the insn matching can be optimized, the if-else + switch case is a bit complicated.

Reply:
like just directly handle insn matching in a large switch case

Reply:
I have switched to a large switch case. And op and func3 is the field name of the RVC instruction.